engine/particles/private/particle_module.cpp

```diff
@@ -1,17 +1,177 @@
+#include <renderer_module.hpp>
 
 ModuleTickOrder ParticleModule::Init(Engine& engine)
 {
-    auto& ecs = engine.GetModule<ECSModule>();
-    auto renderer = engine.GetModule<RendererModule>().GetRenderer();
-    _particleInterface = std::make_unique<ParticleInterface>(renderer->GetContext(), ecs);
+    _physics = &engine.GetModule<PhysicsModule>();
+    _context = engine.GetModule<RendererModule>().GetRenderer()->GetContext();
+    _ecs = &engine.GetModule<ECSModule>();
+
+    return ModuleTickOrder::ePostRender;
+}
+
+void ParticleModule::Tick(MAYBE_UNUSED Engine& engine)
+{
+    const auto emitterView = _ecs->GetRegistry().view<EmitterComponent>();
+    for (const auto entity : emitterView)
+    {
+        if (_ecs->GetRegistry().all_of<RigidbodyComponent>(entity))
+        {
+            const auto& rb = _ecs->GetRegistry().get<RigidbodyComponent>(entity);
+            if (_physics->bodyInterface->GetMotionType(rb.bodyID) != JPH::EMotionType::Static)
+            {
+                _ecs->GetRegistry().emplace_or_replace<ActiveEmitterTag>(entity);
+            }
+        }
+    }
+
+    const auto activeView = _ecs->GetRegistry().view<EmitterComponent, ActiveEmitterTag>();
+    for (const auto entity : activeView)
+    {
+        auto& emitter = _ecs->GetRegistry().get<EmitterComponent>(entity);
+
+        // first remove active tags from inactive emitters and continue
+        if (emitter.emitOnce)
+        {
+            _ecs->GetRegistry().remove<EmitterComponent>(entity);
+            _ecs->GetRegistry().remove<ActiveEmitterTag>(entity);
+            continue;
+        }
+        if (_ecs->GetRegistry().all_of<RigidbodyComponent>(entity))
+        {
+            const auto& rb = _ecs->GetRegistry().get<RigidbodyComponent>(entity);
+
+            // TODO: this is temporary to avoid when an entity doesn't have a WorldMatrixComponent. Would there be a way to do this differently?
+            const auto joltTranslation = _physics->bodyInterface->GetWorldTransform(rb.bodyID).GetTranslation();
+            emitter.emitter.position = glm::vec3(joltTranslation.GetX(), joltTranslation.GetY(), joltTranslation.GetZ());
+
+            if (_physics->bodyInterface->GetMotionType(rb.bodyID) != JPH::EMotionType::Dynamic)
+            {
+                _ecs->GetRegistry().remove<ActiveEmitterTag>(entity);
+                continue;
+            }
+            if (_physics->bodyInterface->GetMotionType(rb.bodyID) != JPH::EMotionType::Static)
+            {
+                JPH::Vec3 rbVelocity = _physics->bodyInterface->GetLinearVelocity(rb.bodyID);
+                emitter.emitter.velocity = -glm::vec3(rbVelocity.GetX(), rbVelocity.GetY(), rbVelocity.GetZ());
+            }
+        }
+
+        // update timers
+        if (emitter.currentEmitDelay < 0.0f)
+        {
+            emitter.currentEmitDelay = emitter.maxEmitDelay;
+        }
+        emitter.currentEmitDelay -= engine.GetModule<TimeModule>().GetDeltatime().count() * 1e-3;
```

Comment:
what is this scaling? ms to s?

Reply:
because if GetDeltaTime returns a std::chrono time point, you could also use a std::chrono::duration_cast\<std::chrono::seconds\>()

Reply:
yeah it's ms to s. I'll try casting